File: src/main/java/net/greenjab/fixedminecraft/registry/item/map_book/MapBookItem.java
package net.greenjab.fixedminecraft.registry.item.map_book;

import net.greenjab.fixedminecraft.registry.registries.ItemRegistry;
import net.greenjab.fixedminecraft.network.MapBookOpenPayload;
import net.greenjab.fixedminecraft.network.MapBookSyncPayload;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.ChatFormatting;
import net.minecraft.core.component.DataComponents;
import net.minecraft.network.chat.Component;
import net.minecraft.network.protocol.Packet;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.tags.BlockTags;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EquipmentSlot;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.MapItem;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.item.component.TooltipDisplay;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.saveddata.maps.MapId;
import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
import net.minecraft.world.phys.Vec3;
import org.jetbrains.annotations.Nullable;
import org.jspecify.annotations.NonNull;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.function.Consumer;

public class MapBookItem extends Item {
    public MapBookItem(Properties settings) {
        super(settings);
    }

    @Override
    public @NonNull InteractionResult useOn(UseOnContext context) {
        BlockState blockState = context.getLevel().getBlockState(context.getClickedPos());

        if (blockState.is(BlockTags.BANNERS)) {
            if (!context.getLevel().isClientSide()) {
                MapStateData mapStateData = this.getNearestMap(context.getItemInHand(), context.getLevel(), context.getClickedPos().getCenter());
                MapItemSavedData mapState = mapStateData.mapState;
                if (mapState != null && !mapState.toggleBanner(context.getLevel(), context.getClickedPos())) {
                    return InteractionResult.FAIL;
                }
            }

            return InteractionResult.SUCCESS;
        } else {
            return super.useOn(context);
        }
    }

    @Override
    public @NonNull InteractionResult use(@NonNull Level world, @NonNull Player user, @NonNull InteractionHand hand) {
        if (!world.isClientSide()) {
            ServerPlayer player = (ServerPlayer) user;
            ItemStack item = user.getItemInHand(hand);
            ItemStack otherHand = hand == InteractionHand.MAIN_HAND ? player.getOffhandItem() : player.getMainHandItem();

            var openMap = true;
            if (getNearestMap(item, world, player.position())==null || otherHand.is(Items.MAP)) {
                if (addNewMapAtPos(item, (ServerLevel)world, player.position(),0)) {
                    if (otherHand.is(Items.MAP) && !player.hasInfiniteMaterials()) {
                        otherHand.shrink(1);
                    }
                    player.level().playSound(
                            null,
                            player,
                            SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT,
                            player.getSoundSource(),
                            1.0f,
                            1.0f
                    );
                    openMap = false;
                }
            } else if (otherHand.is(Items.SHEARS)) {
                if (removeMapAtPos(item, (ServerLevel)world, player.position(), player)) {
                    otherHand.hurtWithoutBreaking(1, player);
                    player.level().playSound(
                            null,
                            player,
                            SoundEvents.SHEEP_SHEAR,
                            player.getSoundSource(),
                            1.0f,
                            1.0f
                    );
                    openMap = false;
                }
            } else if (otherHand.is(Items.FILLED_MAP)) {
                if (addNewMapID(item, otherHand, (ServerLevel)world)) {
                    if (!player.hasInfiniteMaterials()) {
                        otherHand.shrink(1);
                    }
                    player.level().playSound(
                            null,
                            player,
                            SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT,
                            player.getSoundSource(),
                            1.0f,
                            1.0f
                    );
                    openMap = false;
                }
            } else {
                ItemStack hasEmtpyMap = getEmptyMap(user);
                if (hasEmtpyMap.is(Items.MAP)) {
                    boolean hotbar = isHotbar(user, hasEmtpyMap);
                    if (addNewMapAtPos(item, (ServerLevel)world, player.position(), hotbar?2:4)) {
                        if (!player.hasInfiniteMaterials()) {
                            hasEmtpyMap.shrink(1);
                        }
                        player.level().playSound(
                                null,
                                player,
                                SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT,
                                player.getSoundSource(),
                                1.0f,
                                1.0f
                        );
                        openMap = false;
                    }
                }
            }

            this.sendMapUpdates(player, item);
            mapBookSync(player, item);
            if (openMap && this.hasMapBookId(item)) {
                getMapBookState(item, world).update();
                mapBookOpen(player, item);
                player.level().playSound(
                        null,
                        player,
                        SoundEvents.BOOK_PAGE_TURN,
                        player.getSoundSource(),
                        1.0f,
                        1.0f
                );
            }
        }
        return InteractionResult.SUCCESS;
    }

    private ItemStack getEmptyMap(Player playerEntity) {
        for (int i = 0;i < playerEntity.getInventory().getContainerSize(); i++) {
            ItemStack item = playerEntity.getInventory().getItem(i);
            if (item.is(Items.MAP)) return item;
        }
        return ItemStack.EMPTY;
    }
    private boolean isHotbar(Player playerEntity, ItemStack stack) {
        for (int i = 0;i < 9; i++) {
            ItemStack item = playerEntity.getInventory().getItem(i);
            if (item == stack) return true;
        }
        return false;
    }

    private void sendMapUpdates(ServerPlayer player, ItemStack item) {
        for (MapStateData mapStateData : getMapStates(item, player.level())) {
            mapStateData.mapState.getHoldingPlayer(player);
            Packet<?> packet  = mapStateData.mapState.getUpdatePacket(mapStateData.id, player);
            if (packet != null) {
                player.connection.send(packet);
            }
        }
    }

    @Override
    public void inventoryTick(@NonNull ItemStack stack, @NonNull ServerLevel world, @NonNull Entity entity, @Nullable EquipmentSlot slot) {
        if (!world.isClientSide()) {
            if (entity instanceof ServerPlayer player) {
                stack.set(DataComponents.REPAIR_COST, 3);
                int id = getMapBookId(stack);
                if (id != -1) {
                    MapBookState mapBookState = MapBookStateManager.INSTANCE.getMapBookState(world.getServer(), id);
                    if (mapBookState!=null) {
                        mapBookState.addPlayer(player);
                    }
                }

                if (!MapBookStateManager.INSTANCE.currentBooks.contains(id)) {
                    MapBookStateManager.INSTANCE.currentBooks.add(id);
                }
                if ((slot==EquipmentSlot.MAINHAND||slot==EquipmentSlot.OFFHAND) || ((Player) entity).getOffhandItem() == stack) {
                    for (MapStateData mapStateData : getMapStates(stack, entity.level())) {
                        mapStateData.mapState.tickCarriedBy(player, stack, null);
                        if (!mapStateData.mapState.locked) {
                            if (this.getDistanceToEdgeOfMap(mapStateData.mapState, entity.position()) < 128.0) {
                                ((MapItem)Items.FILLED_MAP ).update(world, entity, mapStateData.mapState);
                            }
                        }
                    }
                    this.sendMapUpdates(player, stack);
                }
            }
        }
    }

    public static ArrayList<MapStateData> getMapStates(ItemStack stack, Level world) {
        ArrayList<MapStateData> list = new ArrayList<>();
        MapBookState mapBookState = getMapBookState(stack, world);
        if (mapBookState != null) {
            for (int i : mapBookState.mapIDs) {
                MapItemSavedData mapState = world.getMapData(new MapId(i));
                if (mapState != null) {
                    if (world.dimension().identifier().toString().contains(mapState.dimension.identifier().toString())) {
                        list.add(new MapStateData(new MapId(i), mapState));
                    }
                }
            }
        }
        return list;
    }

    private static MapBookState getMapBookState(ItemStack stack, Level world) {
        int id = getMapBookId(stack);
        if (id == -1) return null;
        if (world.isClientSide()) {
            return MapBookStateManager.INSTANCE.getClientMapBookState(id);
        } else if (world.getServer() != null) {
            return MapBookStateManager.INSTANCE.getMapBookState(world.getServer(), id);
        }
        return null;
    }

    public MapStateData getNearestMap(ItemStack stack, Level world, Vec3 pos) {
        double nearestDistance = Double.MAX_VALUE;
        byte nearestScale = 127;
        MapStateData nearestMap = null;
        Iterator<MapStateData> mapStates = getMapStates(stack, world).iterator();
        while (true) {
            MapStateData mapStateData;
            double distance;
            boolean roughlyEqual;
            do {
                do {
                    if (!mapStates.hasNext()) {
                        return nearestMap;
                    }

                    mapStateData = mapStates.next();
                    distance = this.getDistanceToEdgeOfMap(mapStateData.mapState, pos);
                    if (distance < 0.0) {
                        distance = -1.0;
                    }

                    roughlyEqual = Math.abs(nearestDistance - distance) < 1.0;
                } while (!(distance < nearestDistance) && !roughlyEqual);
            } while (roughlyEqual && (!(distance < 0.0) || mapStateData.mapState.scale >= nearestScale) && (!(distance > 0.0) || mapStateData.mapState.scale <= nearestScale));

            nearestDistance = distance;
            nearestScale = mapStateData.mapState.scale;
            nearestMap = mapStateData;
        }
    }

    private double getDistanceToEdgeOfMap(MapItemSavedData mapState, Vec3 pos) {
        return Math.max(
                Math.abs(pos.x - mapState.centerX),
                Math.abs(pos.z - mapState.centerZ)
        ) - (64 * Math.pow(2, mapState.scale));
    }

    private boolean hasMapBookId(ItemStack stack) {
        return stack.has(DataComponents.MAP_ID);
    }

    public static int getMapBookId(ItemStack stack) {
        MapId mapIdComponent = stack.getOrDefault(DataComponents.MAP_ID, new MapId(-1));
        return mapIdComponent.id();
    }

    private int allocateMapBookId(MinecraftServer server) {

        MapBookIdCountsState counts = server.getDataStorage().computeIfAbsent(
                MapBookIdCountsState.persistentStateType
        );
        int i = counts.get();
        MapBookStateManager.INSTANCE.putMapBookState(server, i, new MapBookState());
        return i;
    }

    private void setMapBookId(ItemStack stack, int id) {
        stack.set(DataComponents.MAP_ID, new MapId(id));
    }

    private int createMapBookState(ItemStack stack, MinecraftServer server) {
        int i = this.allocateMapBookId(server);
        this.setMapBookId(stack, i);
        return i;
    }

    private MapBookState getOrCreateMapBookState(ItemStack stack, MinecraftServer server) {
        int id = getMapBookId(stack);
        MapBookState state = id == -1 ? null : MapBookStateManager.INSTANCE.getMapBookState(server, id);
        if (state != null) {
            return state;
        } else {
            int i = this.createMapBookState(stack, server);
            return MapBookStateManager.INSTANCE.getMapBookState(server, i);
        }
    }

    private boolean addNewMapAtPos(ItemStack item, ServerLevel world, Vec3 pos, int scale) {
        MapBookState state = this.getOrCreateMapBookState(item, world.getServer());
        MapStateData nearestState = this.getNearestMap(item, world, pos);
        if (nearestState != null && nearestState.mapState.scale <= scale
            && !(this.getDistanceToEdgeOfMap(nearestState.mapState, pos) > 0.0)) {
            return false;
        } else {
            ItemStack newMap = MapItem.create(
                    world,
                    (int)Math.floor(pos.x), (int)Math.floor(pos.z), (byte)scale, true, false
            );
            state.addMapID(newMap.get(DataComponents.MAP_ID).id());
            return true;
        }
    }

    private boolean removeMapAtPos(ItemStack item, ServerLevel world, Vec3 pos, ServerPlayer player) {
        MapBookState state = this.getOrCreateMapBookState(item, world.getServer());
        if (state.mapIDs.isEmpty()) return false;
        if (getMapStates(item, world).size()<2) return false;
        MapStateData nearestState = this.getNearestMap(item, world, pos);
        if (nearestState == null) return false;
        if (this.getDistanceToEdgeOfMap(nearestState.mapState, pos) > 0.0) return false;
        if (state.removeMapID(nearestState.id.id())) {
            ItemStack itemStack = new ItemStack(Items.FILLED_MAP);
            itemStack.set(DataComponents.MAP_ID, nearestState.id);
            if (!player.getInventory().add(itemStack)) {
                player.drop(itemStack, true);
            }
        }
        return true;
    }

    private boolean addNewMapID(ItemStack item, ItemStack filledmap, ServerLevel world) {
        MapId mapId = filledmap.get(DataComponents.MAP_ID);
        MapItemSavedData mapState = world.getMapData(mapId);
        MapBookState state = this.getOrCreateMapBookState(item, world.getServer());
        if (state != null && mapState != null && !mapState.locked) {
            if (!state.mapIDs.contains(mapId.id())) {
                state.addMapID(mapId.id());
                return true;
            }
        }
        return false;
    }

    @Override
    public @NonNull Component getName(@NonNull ItemStack stack) {
        if (!this.hasMapBookId(stack)) {
            if (stack.has(ItemRegistry.MAP_BOOK_ADDITIONS)) {
                return Component.translatable("item.fixedminecraft.map_book_new");
            } else {
                return Component.translatable("item.fixedminecraft.map_book_empty");
            }
        } else {
            return super.getName(stack);
        }
    }

    @Override
    public void appendHoverText(ItemStack stack, Item.@NonNull TooltipContext context, @NonNull TooltipDisplay displayComponent, @NonNull Consumer<Component> textConsumer, @NonNull TooltipFlag type) {
        var mapsCount =
                stack.getOrDefault(ItemRegistry.MAP_BOOK_ADDITIONS, MapBookAdditionsComponent.DEFAULT).additions().size();
        int id = getMapBookId(stack);
        if (id != -1) {
            // append tooltip is client-based, so its safe to get the client MapBookState
            MapBookState mapBookState = MapBookStateManager.INSTANCE.getClientMapBookState(id);

            if (mapBookState != null) {
                mapsCount += mapBookState.mapIDs.size();
            }

            textConsumer.accept(Component.translatable("item.fixedminecraft.map_book_id", (id + 1)).withStyle(ChatFormatting.GRAY));
        }

        if (mapsCount > 0) {
            textConsumer.accept(Component.translatable("item.fixedminecraft.map_book_maps", mapsCount).withStyle(ChatFormatting.GRAY));
        }
    }

    private void applyAdditions(ItemStack stack, ServerLevel world) {
        MapBookAdditionsComponent additionsComponent = stack.getOrDefault(ItemRegistry.MAP_BOOK_ADDITIONS, MapBookAdditionsComponent.DEFAULT) ;
        stack.remove(ItemRegistry.MAP_BOOK_ADDITIONS);

        List<Integer> additions = additionsComponent.additions();
        if (!additions.isEmpty()) {
            MapBookState state = this.getOrCreateMapBookState(stack, world.getServer());

            for (int id : additions) {
                if (id != -1) {
                    MapItemSavedData newState = world.getMapData(new MapId(id));

                    state.mapIDs.removeIf( existingID -> {
                        MapItemSavedData existingState = world.getMapData(new MapId(existingID));
                        if (existingState == null) return true;
                        assert newState != null;
                        return (mapsAreSameLocation(newState, existingState));
                    });
                    state.addMapID(id);
                }
            }
        }
    }

    private static boolean mapsAreSameLocation(MapItemSavedData mapA, MapItemSavedData mapB) {
        return mapA.scale == mapB.scale && mapA.centerX == mapB.centerX && mapA.centerZ == mapB.centerZ;
    }

    @Override
    public void onCraftedBy(@NonNull ItemStack stack, @NonNull Player player) {
        super.onCraftedBy(stack, player);
        if (player instanceof ServerPlayer serverPlayerEntity)
            mapBookSync(serverPlayerEntity, stack);
    }

    @Override
    public void onCraftedPostProcess(@NonNull ItemStack stack, Level world) {
        if (!world.isClientSide()) {
            applyAdditions(stack, (ServerLevel)world);
        }
    }

    private void mapBookOpen(ServerPlayer player, ItemStack itemStack) {
        ServerPlayNetworking.send(player, new MapBookOpenPayload(itemStack));
    }

    private void mapBookSync(ServerPlayer player, ItemStack itemStack) {
        MapBookSyncPayload payload = MapBookSyncPayload.of(player, itemStack);
        if (payload != null) {
            ServerPlayNetworking.send(player, payload);
        }
    }
}

File: src/main/kotlin/net/greenjab/fixedminecraft/registry/item/map_book/MapBookItem.kt
package net.greenjab.fixedminecraft.registry.item.map_book

import net.greenjab.fixedminecraft.registry.ItemRegistry
import net.greenjab.fixedminecraft.network.MapBookOpenPayload
import net.greenjab.fixedminecraft.network.MapBookSyncPayload
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking

import net.minecraft.component.DataComponentTypes
import net.minecraft.component.type.MapIdComponent
import net.minecraft.entity.Entity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.FilledMapItem
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.ItemUsageContext
import net.minecraft.item.Items
import net.minecraft.item.map.MapState
import net.minecraft.item.tooltip.TooltipType
import net.minecraft.network.packet.Packet
import net.minecraft.registry.tag.BlockTags
import net.minecraft.server.MinecraftServer
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.server.world.ServerWorld
import net.minecraft.sound.SoundEvents
import net.minecraft.text.Text
import net.minecraft.util.ActionResult
import net.minecraft.util.Formatting
import net.minecraft.util.Hand
import net.minecraft.util.math.Vec3d
import net.minecraft.world.World
import java.util.Objects
import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.max

class MapBookItem(settings: Settings?) : Item(settings) {

    override fun useOnBlock(context: ItemUsageContext): ActionResult {
        val blockState = context.world.getBlockState(context.blockPos)

        if (blockState.isIn(BlockTags.BANNERS)) {
            if (!context.world.isClient) {
                val mapStateData = this.getNearestMap(context.stack, context.world, context.blockPos.toCenterPos())
                val mapState: MapState? = mapStateData?.mapState
                if (mapState != null && !mapState.addBanner(context.world, context.blockPos)) {
                    return ActionResult.FAIL
                }
            }

            return ActionResult.SUCCESS
        } else {
            return super.useOnBlock(context)
        }
    }

    /*override fun use(world: World, user: PlayerEntity, hand: Hand): ActionResult {
        if (world != null && !world.isClient()) {
            val player = user as ServerPlayerEntity
            val item = user.getStackInHand(hand)
            val otherHand = if (hand == Hand.MAIN_HAND) player.offHandStack else player.mainHandStack
            var openMap = true
            if (otherHand.isOf(Items.PAPER)) {
                if (this.addNewMapAtPos(item, world as ServerWorld, player.pos, 0)) {
                    if (!player.abilities.creativeMode) {
                        otherHand.decrement(1)
                    }

                    player.world.playSoundFromEntity(
                        null,
                        player,
                        SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT,
                        player.soundCategory,
                        1.0f,
                        1.0f
                    )
                    openMap = false
                }
            }
            this.sendMapUpdates(player, item)
            mapBookSync(player, item)
            if (openMap && this.hasMapBookId(item)) {
                mapBookOpen(player, item)
            }
        }

        return super.use(world, user, hand)
    }*/

    override fun use(world: World?, user: PlayerEntity?, hand: Hand?): ActionResult {
        if (world != null && !world.isClient()) {
            val player = user as ServerPlayerEntity
            val item = user.getStackInHand(hand)
            val otherHand = if (hand == Hand.MAIN_HAND) player.offHandStack else player.mainHandStack

            var openMap = true
            if (getNearestMap(item, world as ServerWorld, player.pos)==null) {
                if (addNewMapAtPos(item, world, player.pos,0)) {
                    player.world.playSoundFromEntity(
                        null,
                        player,
                        SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT,
                        player.soundCategory,
                        1.0f,
                        1.0f
                    )
                    openMap = false
                }
            } else if (otherHand.isOf(Items.PAPER)) {
                if (addNewMapAtPos(item, world, player.pos, 0)) {
                    if (!player.abilities.creativeMode) {
                        otherHand.decrement(1)
                    }
                    player.world.playSoundFromEntity(
                        null,
                        player,
                        SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT,
                        player.soundCategory,
                        1.0f,
                        1.0f
                    )
                    openMap = false
                }
            } else if (otherHand.isOf(Items.FILLED_MAP)) {
                if (addNewMapID(item, otherHand, world)) {
                    if (!player.abilities.creativeMode) {
                        otherHand.decrement(1)
                    }
                    player.world.playSoundFromEntity(
                        null,
                        player,
                        SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT,
                        player.soundCategory,
                        1.0f,
                        1.0f
                    )
                    openMap = false
                }
            } else {
                val hasPaper = getPaper(user)
                if (hasPaper.isOf(Items.PAPER)) {
                    if (addNewMapAtPos(item, world, player.pos, 0)) {
                        if (!player.abilities.creativeMode) {
                            hasPaper.decrement(1)
                        }
                        player.world.playSoundFromEntity(
                            null,
                            player,
                            SoundEvents.UI_CARTOGRAPHY_TABLE_TAKE_RESULT,
                            player.soundCategory,
                            1.0f,
                            1.0f
                        )
                        openMap = false
                    }
                }
            }

            this.sendMapUpdates(player, item)
            mapBookSync(player, item)
            if (openMap && this.hasMapBookId(item)) {
                this.getMapBookState(item, world)?.update()
                mapBookOpen(player, item)
            }
        }
        return ActionResult.SUCCESS
        //return super.use(world, user, hand)
    }

    private fun getPaper(playerEntity: PlayerEntity): ItemStack {
        for (i in 0 until playerEntity.inventory.size()) {
            val item = playerEntity.inventory.getStack(i)
            if (item.isOf(Items.PAPER)) return item
        }
        return ItemStack.EMPTY
    }

    private fun sendMapUpdates(player: ServerPlayerEntity, item: ItemStack) {
        for (mapStateData in this.getMapStates(item, player.world)) {
            mapStateData.mapState.getPlayerSyncData(player)
            val packet: Packet<*>? = mapStateData.mapState.getPlayerMarkerPacket(mapStateData.id, player)
            if (packet != null) {
                player.networkHandler.sendPacket(packet)
            }
        }
    }

    override fun inventoryTick(stack: ItemStack?, world: World?, entity: Entity?, slot: Int, selected: Boolean) {
        if (world != null && !world.isClient()) {
            if (stack != null && entity is PlayerEntity) {
                if (selected || entity.offHandStack == stack) {
                    // this.applyAdditions(stack, (ServerWorld)world);

                    for (mapStateData in this.getMapStates(stack, entity.getWorld())) {
                        mapStateData.mapState.update(entity as PlayerEntity?, stack)
                        if (!mapStateData.mapState.locked) {
                            if (this.getDistanceToEdgeOfMap(mapStateData.mapState, entity.getPos()) < 128.0) {
                                (Items.FILLED_MAP as FilledMapItem).updateColors(world, entity, mapStateData.mapState)
                            }
                        }
                    }
                    val id = getMapBookId(stack)
                    world.server?.let {
                        if (id != null) {
                            MapBookStateManager.INSTANCE.getMapBookState(it, id)?.addPlayer(entity)
                        }
                    }
                    if (!MapBookStateManager.INSTANCE.currentBooks.contains(id)) id?.let { MapBookStateManager.INSTANCE.currentBooks.add(it) }
                    this.sendMapUpdates(entity as ServerPlayerEntity, stack)
                    //mapBookSync(entity, stack)
                }
            }
        }
    }

    fun getMapStates(stack: ItemStack, world: World): ArrayList<MapStateData> {
        val list = ArrayList<MapStateData>()
        val mapBookState = getMapBookState(stack, world)

        if (mapBookState != null) {
            for (i in mapBookState.mapIDs) {
                val mapState = world.getMapState(MapIdComponent(i))
                if (mapState != null) {
                    list.add(MapStateData(MapIdComponent(i), mapState))
                }
            }
        }
        return list
    }

    private fun getMapBookState(stack: ItemStack, world: World): MapBookState? {
        val id = getMapBookId(stack) ?: return null

        if (world.isClient) {
            return MapBookStateManager.INSTANCE.getClientMapBookState(id)
        } else if (world.server != null) {
            return MapBookStateManager.INSTANCE.getMapBookState(world.server!!, id)
        }
        return null
    }

    fun getNearestMap(stack: ItemStack, world: World, pos: Vec3d): MapStateData? {
        var nearestDistance = Double.MAX_VALUE
        var nearestScale: Byte = 127
        var nearestMap: MapStateData? = null

        val mapStates: Iterator<MapStateData> = getMapStates(stack, world).iterator()

        while (true) {
            var mapStateData: MapStateData
            var distance: Double
            var roughlyEqual: Boolean
            do {
                do {
                    if (!mapStates.hasNext()) {
                        return nearestMap
                    }

                    mapStateData = mapStates.next()
                    distance = this.getDistanceToEdgeOfMap(mapStateData.mapState, pos)
                    if (distance < 0.0) {
                        distance = -1.0
                    }

                    roughlyEqual = abs(nearestDistance - distance) < 1.0
                } while (!(distance < nearestDistance) && !roughlyEqual)
            } while (roughlyEqual && (!(distance < 0.0) || mapStateData.mapState.scale >= nearestScale) && (!(distance > 0.0) || mapStateData.mapState.scale <= nearestScale))

            nearestDistance = distance
            nearestScale = mapStateData.mapState.scale
            nearestMap = mapStateData
        }
    }

    private fun getDistanceToEdgeOfMap(mapState: MapState, pos: Vec3d): Double {
        return max(
            abs(pos.x - mapState.centerX.toDouble()),
            abs(pos.z - mapState.centerZ.toDouble())
        ) - (64 * (1 shl mapState.scale.toInt())).toDouble()
    }

    private fun hasMapBookId(stack: ItemStack): Boolean {
        return stack.contains(DataComponentTypes.MAP_ID)
    }

    fun getMapBookId(stack: ItemStack): Int? {
        val mapIdComponent = stack.getOrDefault(DataComponentTypes.MAP_ID, null)
        return mapIdComponent?.id()
    }

    private fun allocateMapBookId(server: MinecraftServer): Int {
        val counts: MapBookIdCountsState = server.overworld.persistentStateManager.getOrCreate(
            MapBookIdCountsState.persistentStateType,
            MapBookIdCountsState.IDCOUNTS_KEY
        )
        val i = counts.nextMapBookId
        MapBookStateManager.INSTANCE.putMapBookState(server, i, MapBookState())
        return i
    }

    private fun setMapBookId(stack: ItemStack, id: Int) {
        stack.set(DataComponentTypes.MAP_ID, MapIdComponent(id))
    }

    private fun createMapBookState(stack: ItemStack, server: MinecraftServer): Int {
        val i = this.allocateMapBookId(server)
        this.setMapBookId(stack, i)
        return i
    }

    private fun getOrCreateMapBookState(stack: ItemStack, server: MinecraftServer): MapBookState? {
        val id = this.getMapBookId(stack)
        val state = if (id == null) null else MapBookStateManager.INSTANCE.getMapBookState(server, id)
        if (state != null) {
            return state
        } else {
            val i = this.createMapBookState(stack, server)
            return MapBookStateManager.INSTANCE.getMapBookState(server, i)
        }
    }

    private fun addNewMapAtPos(item: ItemStack, world: ServerWorld, pos: Vec3d, scale: Int): Boolean {
        val state = this.getOrCreateMapBookState(item, world.server)
        val nearestState = this.getNearestMap(item, world, pos)
        if (nearestState != null && nearestState.mapState.scale <= scale && !(this.getDistanceToEdgeOfMap(
                nearestState.mapState,
                pos
            ) > 0.0)
        ) {
            return false
        } else {
            val newMap = FilledMapItem.createMap(
                world,
                floor(pos.x).toInt(), floor(pos.z).toInt(), scale.toByte(), true, false
            )
            Objects.requireNonNull(newMap.get(DataComponentTypes.MAP_ID))?.let { state!!.addMapID(it.id) }
            return true
        }
    }

    private fun addNewMapID(item: ItemStack, filledmap: ItemStack, world: ServerWorld): Boolean {
        val mapId = filledmap.get(DataComponentTypes.MAP_ID) ?: return false
        val state = this.getOrCreateMapBookState(item, world.server)
        if (state != null) {
            if (!state.mapIDs.contains(mapId.id)) {
                Objects.requireNonNull(filledmap.get(DataComponentTypes.MAP_ID))?.let { state.addMapID(it.id) }
                return true
            }
        }
        return false
    }

    override fun getName(stack: ItemStack?): Text {
        return if (stack != null && !this.hasMapBookId(stack)) {
            if (stack.contains(ItemRegistry.MAP_BOOK_ADDITIONS)) {
                Text.translatable("item.fixedminecraft.map_book_new")
            } else {
                Text.translatable("item.fixedminecraft.map_book_empty")
            }
        } else {
            super.getName(stack)
        }
    }

    override fun appendTooltip(stack: ItemStack, context: TooltipContext, tooltip: MutableList<Text>, type: TooltipType) {
        var mapsCount =
            stack.getOrDefault(ItemRegistry.MAP_BOOK_ADDITIONS, MapBookAdditionsComponent.DEFAULT).additions.size
        val id = this.getMapBookId(stack)
        if (id != null) {
            // append tooltip is client-based, so its safe to get the client MapBookState
            val mapBookState = MapBookStateManager.INSTANCE.getClientMapBookState(id)

            if (mapBookState != null) {
                mapsCount += mapBookState.mapIDs.size
            }

            tooltip.add(Text.translatable("item.fixedminecraft.map_book_id", (id + 1).toString()).formatted(Formatting.GRAY))
        }

        if (mapsCount > 0) {
            tooltip.add(Text.translatable("item.fixedminecraft.map_book_maps", mapsCount.toString()).formatted(Formatting.GRAY))
        }
    }

    fun setAdditions(stack: ItemStack, additions: List<Int>) {
        stack.set<MapBookAdditionsComponent>(ItemRegistry.MAP_BOOK_ADDITIONS,
            MapBookAdditionsComponent(additions)
        )
    }

    private fun applyAdditions(stack: ItemStack, world: ServerWorld) {
        val additionsComponent = stack.getOrDefault<MapBookAdditionsComponent>(ItemRegistry.MAP_BOOK_ADDITIONS, null) ?: return
        stack.remove<Any>(ItemRegistry.MAP_BOOK_ADDITIONS)

        val additions = additionsComponent.additions
        if (additions.isNotEmpty()) {
            val state = this.getOrCreateMapBookState(stack, world.server)

            for (id in additions) {
                val newState = world.getMapState(id?.let { MapIdComponent(it) }) ?: continue

                state!!.mapIDs.removeIf { existingID: Int? ->
                    val existingState = world.getMapState(MapIdComponent(existingID!!))
                    existingState == null || mapsAreSameLocation(newState, existingState)
                }

                if (id != null) {
                    state.addMapID(id)
                }
            }
        }
    }

    fun hasInvalidAdditions(stack: ItemStack, world: World, additions: List<Int>): Boolean {
        val mapBookState = getMapBookState(stack, world)

        for (i in additions.indices) {
            val additionA = additions[i]
            if (mapBookState != null && mapBookState.mapIDs.contains(additionA)) {
                return true
            }

            val mapA = world.getMapState(MapIdComponent(additionA)) ?: return true

            for (j in i + 1 until additions.size) {
                val additionB = additions[j]
                if (additionA == additionB) {
                    return true
                }

                val mapB = world.getMapState(MapIdComponent(additionB))
                if (mapB == null || mapsAreSameLocation(mapA, mapB)) {
                    return true
                }
            }
        }

        return false
    }

    private fun mapsAreSameLocation(mapA: MapState, mapB: MapState): Boolean {
        return mapA.scale == mapB.scale && mapA.centerX == mapB.centerX && mapA.centerZ == mapB.centerZ
    }

    override fun onCraftByPlayer(stack: ItemStack, world: World, player: PlayerEntity) {
        super.onCraftByPlayer(stack, world, player)
        if (!world.isClient) {
            mapBookSync(player as ServerPlayerEntity, stack)
        }
    }

    override fun onCraft(stack: ItemStack, world: World) {
        if (!world.isClient) {
            applyAdditions(stack, world as ServerWorld)
        }
    }

    private fun mapBookOpen(player: ServerPlayerEntity, itemStack: ItemStack) {
        ServerPlayNetworking.send(player, MapBookOpenPayload(itemStack))
    }

    private fun mapBookSync(player: ServerPlayerEntity, itemStack: ItemStack) {
        val payload: MapBookSyncPayload? = MapBookSyncPayload.of(player, itemStack)
        if (payload != null) {
            ServerPlayNetworking.send(player, payload)
        }
    }
}
